lib/cos.js

```diff
@@ -0,0 +1,134 @@
+
+exports.initialize = function initializeDataSource(dataSource, callback) {
+	//console.log(dataSource.settings);
+	dataSource.connector = new COSConnector(dataSource.settings);
+	process.nextTick(function () {
+		callback && callback();
+	});
+};
+var Connector = require('loopback-connector').Connector;
+
+require('util').inherits(COSConnector, Connector);
+
+const IBM = require('ibm-cos-sdk');
+
+exports.COSConnector = COSConnector;
+
+function COSConnector(dataSourceProps) {
+
+	this.endpoint = dataSourceProps.endpoint;
+	this.apiKeyId = dataSourceProps.apiKeyId;
+	this.ibmAuthEndpoint = 'https://iam.ng.bluemix.net/oidc/token';
+	this.serviceInstanceId = dataSourceProps.serviceInstanceId;
+	this.bucket = dataSourceProps.bucket;
+	this.response = [];
+	this._models = {};
+
+	this.config = {
+		endpoint: this.endpoint,
+		apiKeyId: this.apiKeyId,
+		ibmAuthEndpoint: this.ibmAuthEndpoint,
+		serviceInstanceId: this.serviceInstanceId,
+	};
+	this.cos = new IBM.S3(this.config);
+
+}
+
+
+//creates entry in bucket
+COSConnector.prototype.create = function create(model, data, callback){
+
+	var key = guid();
+	var dataStr = JSON.stringify(data); 
+
+	return this.cos.putObject({
+		'Bucket': this.bucket, 
```

Comment:
I think it would be better to use `model` (it's a string containing the model name) as the `Bucket`. That way the app can have multiple models attached to the same datasource, as is usual in other connectors.

What do you think?

Alternatively, maybe you can encode the model name in `key`?

```js
const key = `${model}:${guid()}`;
```

Reply:
Not sure exactly what you mean by this point. The user creates a bucket before they use this connector and feed this bucket name in to the connector.
The `key` is the name of the object that is POSTed to the bucket.

Reply:
In LoopBack, multiple models can be backed by the same datasource (connector instance).

In the SQL database world, an LB app typically has:
 - One datasource (e.g. 'db') configured to connector to a database server.
 - Multiple model classes (e.g. `Category` and `Product`) backed by this same datasource (`db`).
 - The connector is mapping models to SQL tables, e.g. `Category` is stored in the table called `CATEGORY`.

Since your COS connector is implementing CRUD-like API similar to what our database connectors provide, I think it should be also capable to persist multiple model classes using the same single datasource.

The reason why this is important:
 - Attaching multiple models to one datasource is an integral part of LoopBack design. I don't think there is any easy way how to prevent users from doing this.
 - In your current design, when there are multiple models attached to the same datasource (COS connector instance), then the data of different models are mixed together. Saving a Category with id 1 will overwrite data of Product with id 1. Getting data of Category with id 1 can return data of a Product instead etc.

> The user creates a bucket before they use this connector and feed this bucket name in to the connector.

I am saying that the bucket name should not be configured at connector/datasource level, but it should be configured at model level, with a reasonable default constructed by the framework using the model name. That's one way how to solve the problem I have described above.

```ts
const bucket = modelName; // all connector methods are receiving modelName
// e.g. bucket is Product
```

Another option is to keep your design when only one bucket is configured at datasource level,  but change the keys to be namespaced using model names.

```ts
const key = `${model}:${guid()}`;
// e.g. key is Product:1234...
```